core/src/test/java/org/apache/brooklyn/core/config/ConfigKeyConstraintTest.java

```diff
@@ -0,0 +1,285 @@
+                .description("Configuration key that must not be null")
+                .constraint(new MatchesEntityDisplayNamePredicate())
+                .build();
+    }
+    public static class EntityWithContextAwareConstraintImpl extends TestEntityImpl implements EntityWithContextAwareConstraint {
+    }
+
+    public static class PolicyWithConfigConstraint extends AbstractPolicy {
+        public static final ConfigKey<Object> NON_NULL_CONFIG = ConfigKeys.builder(Object.class)
+                .name("test.policy.non-null")
+                .description("Configuration key that must not be null")
+                .constraint(Predicates.notNull())
+                .build();
+    }
+
+    public static class EnricherWithConfigConstraint extends AbstractEnricher {
+        public static final ConfigKey<String> PATTERN = ConfigKeys.builder(String.class)
+                .name("test.enricher.regex")
+                .description("Must match a valid IPv4 address")
+                .constraint(Predicates.containsPattern(Networking.VALID_IP_ADDRESS_REGEX))
+                .build();
+    }
+
+    private static class MatchesEntityDisplayNamePredicate implements BrooklynObjectPredicate<String> {
+        @Override
+        public boolean apply(String input) {
+            return false;
+        }
+
+        @Override
+        public boolean apply(String input, BrooklynObject context) {
+            return context != null && context.getDisplayName().equals(input);
+        }
+    }
+
+    // ----------- Tests -----------------------------------------------------------------------------------------------
+
+    @Test
+    public void testExceptionWhenEntityHasNullConfig() {
+        try {
+            app.createAndManageChild(EntitySpec.create(EntityWithNonNullConstraint.class));
+            fail("Expected exception when managing entity with missing config");
+        } catch (Exception e) {
+            Throwable t = Exceptions.getFirstThrowableOfType(e, ConstraintViolationException.class);
+            assertNotNull(t);
+        }
+    }
+
+    @Test
+    public void testNoExceptionWhenEntityHasValueForRequiredConfig() {
+        app.createAndManageChild(EntitySpec.create(EntityWithNonNullConstraint.class)
+                .configure(EntityWithNonNullConstraint.NON_NULL_CONFIG, new Object()));
+    }
+
+    @Test
+    public void testNoExceptionWhenDefaultValueIsValid() {
+        app.createAndManageChild(EntitySpec.create(EntityRequiringConfigKeyInRange.class));
+    }
+
+    @Test
+    public void testExceptionWhenSubclassSetsInvalidDefaultValue() {
+        try {
+            app.createAndManageChild(EntitySpec.create(EntityProvidingDefaultValueForConfigKeyInRange.class));
+            fail("Expected exception when managing entity setting invalid default value");
+        } catch (Exception e) {
+            Throwable t = Exceptions.getFirstThrowableOfType(e, ConstraintViolationException.class);
+            assertNotNull(t);
```

Comment:
personal preference for something like:

    if (Exceptions.getFirstThrowableOfType(e, ConstraintViolationException.class) == null) {
        throw e;
    }

Otherwise we lose the underlying exception - which is normally fine, but is a real pain when a test fails non-deterministically in the asfbot!